modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java

```diff
@@ -265,19 +269,47 @@ protected int processStartTag() throws Exception {
 
 		jspWriter.write("</li>");
 
-		int totalPages = (int)Math.ceil(_totalItems / _activeDelta);
+		int totalPages = (int)Math.ceil(
+			(double)_totalItems / (double)_activeDelta);
+
+		if (_ellipsisBuffer != 0) {
+			int activeIndex = _activePage - 1;
+
+			int[] totalItems = new int[totalPages];
+
+			for (int i = 0; i < totalPages; i++) {
+				totalItems[i] = i + 1;
+			}
+
+			int lastIndex = totalItems.length - 1;
+			int leftBufferEnd = activeIndex - _ellipsisBuffer;
+			int rightBufferStart = activeIndex + _ellipsisBuffer + 1;
 
-		int ellipsisCount = _activePage + _ellipsisBuffer;
+			int[] leftBuffer = Arrays.copyOfRange(
+				totalItems, 1, Math.max(leftBufferEnd, 1));
+
+			if (rightBufferStart > lastIndex) {
+				rightBufferStart = lastIndex;
+			}
+
+			int[] rightBuffer = Arrays.copyOfRange(
```

Comment:
Typically we would define this variables just before using them to see more clearly the logic.

Reply:
Sure, let me reorganize that.

Reply:
Should look better in https://github.com/liferay-frontend/liferay-portal/pull/1501/commits/ae06adc4a85b942d6add42f8202812a7d3fd4d5a